pkg/broker/handler/handler.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"context"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors"
+	"go.uber.org/zap"
+	"knative.dev/eventing/pkg/logging"
+)
+
+// Handler pulls Pubsub messages as events and process them
+// with chain of processors.
+type Handler struct {
+	// PubsubEvents is the CloudEvents Pubsub protocol to pull
+	// messages as events.
+	PubsubEvents *pubsub.Protocol
+
+	// Processor is the processor to process events.
+	Processor processors.Processor
+
+	// Timeout is the timeout for processing each individual event.
+	Timeout time.Duration
+
+	// cancel is function to stop pulling messages.
+	cancel context.CancelFunc
+}
+
+// Start starts the handler.
+// done func will be called if the pubsub inbound is closed.
+func (h *Handler) Start(ctx context.Context, done func(error)) {
+	ctx, h.cancel = context.WithCancel(ctx)
+	go func() {
+		done(h.PubsubEvents.OpenInbound(ctx))
+	}()
+
+	go h.handle(ctx)
+}
+
+// Stop stops the handlers.
+func (h *Handler) Stop() {
+	h.cancel()
+}
+
+func (h *Handler) handle(ctx context.Context) {
+	for {
+		// TODO(yolocs): we need to update dep for fix:
+		// https://github.com/cloudevents/sdk-go/pull/430
+		msg, err := h.PubsubEvents.Receive(ctx)
+		if err != nil {
+			logging.FromContext(ctx).Error("failed to receive the next message from Pubsub", zap.Error(err))
+			continue
+		}
+
+		event, err := binding.ToEvent(ctx, msg)
+		if err != nil {
+			logging.FromContext(ctx).Error("failed to convert received message to an event", zap.Error(err))
+			continue
+		}
+
+		// Don't block new events.
+		go func() {
```

Comment:
Is this really necessary? Are we sure that handle() isn't already called in its own goroutine?

Reply:
I believe so. Looking at the pubsub protocol code, even though the pubsub client pulls messages with multiple goroutines, they all end up in the same channel. The `Receive` func simply reads from the channel.

Reply:
`handle()` is called in its goroutine, but here in `handle`, we also want to parallelize processing of events, so I also think the goroutine is needed here.

However, I was wondering if we should control the total number of goroutines here. Suppose we can pull events from pubsub very quickly while the `Processors` are slow, we may end up with OOM.

Reply:
Good point. Let me add that.